doc: JSON schema for cockpit `manifest.json`

Venefilyn opened the pull request:
Our `manifest.json` is currently only documented through our docs, and
in there it can be quite confusing. We tend to instead copy from other
repos or pages and I assume other projects do too. But modifying it can
be tedious.

Setting this JSON schema as a `$schema` property in your `manifest.json`
file will make all properties validated against.

For usage within manifest files the schema will be published under
Antora as an attachment and will therefore be versioned. This can be
accessed using the following for example:
- https://docs.cockpit-project.org/cockpit-guide/latest/_attachments/schema.json
- https://docs.cockpit-project.org/cockpit-guide/363/_attachments/schema.json

Assisted-by: AI

Venefilyn committed:
doc: JSON schema for cockpit `manifest.json`

Our `manifest.json` is currently only documented through our docs, and
in there it can be quite confusing. We tend to instead copy from other
repos or pages and I assume other projects do too. But modifying it can
be tedious.

Setting this JSON schema as a `$schema` property in your `manifest.json`
file will make all properties validated against.

For usage within manifest files the schema will be published under
Antora as an attachment and will therefore be versioned. This can be
accessed using the following for example:
- https://docs.cockpit-project.org/cockpit-guide/latest/_attachments/schema.json
- https://docs.cockpit-project.org/cockpit-guide/363/_attachments/schema.json

Assisted-by: AI
Signed-off-by: Freya Gustavsson <freya@venefilyn.se>

Venefilyn commented:
Then we can have it on the website as a recommendation:
```json
{
    "$schema": "https://docs.cockpit-project.org/cockpit-guide/latest/_attachments/schema.json"
    ...
}
```

And to test locally before it's published seems tricky, but you might be able to do something like this. It might also be possible to "embed" by referring a URL to a local file instead and then embed, but not yet tested.

`cockpit/pkg/networkmanager/manifest.json`
```json
{
    "$schema": "../../doc/modules/ROOT/attachments/schema.json",
    "name": "network",
    "conditions": [
        {"path-exists": "/usr/share/dbus-1/system.d/org.freedesktop.NetworkManager.conf"}
    ],
    "menu": {
        "index": {
            "label": "Networking",
            "order": 40,
            "docs": [
                {
                    "label": "Managing networking bonds",
                    "url": "https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/configuring_and_managing_networking/configuring-network-bonding_configuring-and-managing-networking#proc_configuring-a-network-bond-by-using-the-rhel-web-console_configuring-network-bonding"
                },
                {
                    "label": "Managing networking teams",
                    "url": "https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/configuring_and_managing_networking/configuring-network-teaming_configuring-and-managing-networking#proc_configuring-a-network-team-by-using-the-rhel-web-console_configuring-network-teaming"
                },
                {
                    "label": "Managing networking bridges",
                    "url": "https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/configuring_and_managing_networking/configuring-a-network-bridge_configuring-and-managing-networking#proc_configuring-a-network-bridge-by-using-the-rhel-web-console_configuring-a-network-bridge"
                },
                {
                    "label": "Managing VLANs",
                    "url": "https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/configuring_and_managing_networking/configuring-vlan-tagging_configuring-and-managing-networking#proc_configuring-vlan-tagging-by-using-the-rhel-web-console_configuring-vlan-tagging"
                },
                {
                    "label": "Managing firewall",
                    "url": "https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/configuring_firewalls_and_packet_filters/using-and-configuring-firewalld_firewall-packet-filters#enabling-services-on-firewall-using-the-web-console_controlling-network-traffic-using-firewalld"
                }
            ],
            "keywords": [
                {
                    "matches": ["network", "interface", "bridge", "vlan", "bond", "team", "port", "mac", "ipv4", "ipv6"]
                },
                {
                    "matches": ["firewall", "firewalld", "zone", "tcp", "udp"],
                    "goto": "/network/firewall"
                }
            ]
        }
    }
}
```

mvollmer commented on doc/guide/static/manifest.schema.json:
```diff
@@ -0,0 +1,256 @@
+      }
+    },
+    "tools": {
+      "description": "todo",
```
"tools" and "dashboard" should be the same as "menu", i.e."#/$defs/menu".

Venefilyn replied:
Ack thanks! Lots of lines so explains why I missed a few things

mvollmer commented:
This seems to still be draft, right?

mvollmer commented:
@Venefilyn, in case you haven't already, please have a look at `pkg/shell/manifests.ts`, which validates those parts of the manifests that are of interest to the Shell.  (I.e., not the "bridges" and "conditions" fields, for example.)

mvollmer commented:
I like this! We should validate our own manifests according to this schema in test/common/static-code.

Venefilyn commented on doc/guide/static/manifest.schema.json:
```diff
@@ -0,0 +1,256 @@
+    "preload": {
+      "type": "array",
+      "description": "A list of identifiers of the components that should be preloaded. Normally, the files of a component are loaded when the user navigates to it for the first time. The files of a preloaded component are loaded immediately after the user logs in, and the initialization code of the component is invoked. \n\nThe value of this field is an array of strings, where each string is one ...
+    },
```
Yeah I'll change it to
```json
"preload": {
  "type": "array",
  "description": "...",
  "items": {
    "type": "string"
  }
},
```

allisonkarlitskaya reviewed (commented):
I know this sounds ridiculous but could we write this file in yaml?  It would be a lot easier to read...

Otherwise: before we commit to this I think I want a firmer idea of where the schema will live.  In particular:
 - we need to run our linters and stuff offline, which implies that the schema needs to live in the cockpit repo in a place where the subprojects can also hit it
 - you seemed to suggest that we'll replace the relative path with a URL during publishing but we don't perform any transformations on our manifest.json files anymore (which was a hard-fought victory).  Will the symlink be broken when the file is installed?  Will we install the schema on the system?
 - will the bridge break if it notices this field and doesn't recognize it?  I guess/hope not?

Venefilyn commented:
> I know this sounds ridiculous but could we write this file in yaml? It would be a lot easier to read...

There aren't really yaml schema validation AFAIK.

> Otherwise: before we commit to this I think I want a firmer idea of where the schema will live. In particular:
> 
>     * we need to run our linters and stuff offline, which implies that the schema needs to live in the cockpit repo in a place where the subprojects can also hit it
> 
>     * you seemed to suggest that we'll replace the relative path with a URL during publishing but we don't perform any transformations on our manifest.json files anymore (which was a hard-fought victory).  Will the symlink be broken when the file is installed?  Will we install the schema on the system?
> 
>     * will the bridge break if it notices this field and doesn't recognize it?  I guess/hope not?

This wouldn't really change the existing syntax whatsoever though. But I also hope that we don't break cockpit if it detects a top level property "$schema" on a `manifest.json`.

* Schema validation is supposed to be done either live as you're editing (as it can also give descriptions when you're editing) or through a tool that reads the schema and validates against it
* One change is suggested to `manifest.json` in which other applications should be allowed to have a `$schema` top level property
* If validation fails, it wouldn't cause any issues currently. But if we change our validation script to use the schema then I suppose yes it would. But having tests to cover the current manifest validation implementation then migrating the validation to use the schema should not cause any tests to fail. It should not cause any breaking changes.

But yeah I don't suggest we throw out our validation we have now for this, but rather recommend to use this schema for your `manifest.json` to have it much easier to know if something is broken or not and what fields exist to use.

Venefilyn commented:
I'm gonna block this on #21515 to make things easier for myself

Venefilyn committed:
workflows: Run JSON file validation

Validated our JSON files against JSON standard while also supporting
`$schema` validation so our `manifest.json` can be verified to match our
manifest JSON schema.

Signed-off-by: Freya Gustavsson <freya@venefilyn.se>

Venefilyn commented:
@mvollmer I've updated this now so that it will be published in Antora under each version, so if you look at https://docs.cockpit-project.org you would be able to access a versioned schema from there, such as `https://docs.cockpit-project.org/cockpit-guide/363/_attachments/schema.json`

Note that these URLs are just examples and not live until this is merged and docs rebuilt. After that the next version with this PR included will have it available

jelly commented on doc/modules/ROOT/attachments/schema.json:
```diff
@@ -0,0 +1,229 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "https://cockpit-project.org/manifest-1/schema",
+  "title": "Cockpit AppStream manifest file validation for Cockpit pages",
```
What's an Cockpit Appstream manifest file? This isnt appstream related at all?

jelly commented on doc/modules/ROOT/attachments/schema.json:
```diff
@@ -0,0 +1,229 @@
+  "type": "object",
+  "properties": {
+    "content-security-policy": {
+      "description": "By default Cockpit serves packages using a strict Content Security Policy, which among other things does not allow inline styles or scripts. This can be overridden on a per-package basis, with this setting. \n\nIf the overridden content security policy does not contain a `default-src`, `connect-src`, `base-uri`, `form-action`, `object-src`, or `block-all-mixed-content` then  ...
```
Documenting the default value is going to end up bitrotting away when we change it. For example `block-all-mixed-content` is already a deprecated CSP setting.

Venefilyn replied:
For the values we ignore we should still list them here. Our docs are outdated too if this is

Venefilyn replied:
Our docs mention we only take from manifest if it doesn't include any of the listed values. Is that wrong?

jelly replied:
Ah, this is just more copy-pasted stuff from the docs...

jelly commented on doc/modules/ROOT/attachments/schema.json:
```diff
@@ -0,0 +1,229 @@
+      "type": "string"
+    },
+    "name": {
+      "description": "An optional string that changes the name of the package. Normally packages derive their name from the directory that they are located in. This field overrides that name",
```
We don't allow spaces here and other rules:

The name of the package must be ASCII alphanumeric, and may contain an
underscore. Names of directories and files in the package must consist
of ASCII alphanumeric along with dash, underscore, dot, and comma. No
spaces are allowed.

Venefilyn replied:
So for package name itself it should be `^[A-Za-z0-9_]+$`?

Venefilyn replied:
Are there limitations for having underscore in beginning or end or multiple in succession?

jelly replied:
I don't think so, "The name of a package is the name of the directory." (this is the default I think)

jelly commented on doc/modules/ROOT/attachments/schema.json:
```diff
@@ -0,0 +1,229 @@
+      "properties": {
+        "cockpit": {
+          "type": "string",
+          "pattern": "^\\d+$"
```
We support point releases:

```
>>> cockpit.__version__
'334.2'
```

This is perfectly valid.

jelly commented on doc/modules/ROOT/attachments/schema.json:
```diff
@@ -0,0 +1,229 @@
+            "type": "string"
+          },
+          "path-not-exists": {
+            "description": "A condition where a specific file path must exist",
```
must not exist.

jelly commented on doc/modules/ROOT/attachments/schema.json:
```diff
@@ -0,0 +1,229 @@
+      "type": "number"
+    },
+    "conditions": {
+      "description": "An optional list of `{\"predicate\": \"value\"}` objects. Cockpit will only consider the package if all conditions are met. Currently supported predicates are `path-exists` and `path-not-exists`. Unknown predicates are ignored. This is preferable to using `priority`, but only available since Cockpit 286",
```
This also supports `{"any": [{"path-exists": "/usr/bin/sos"}, {"path-exists": "/usr/sbin/sos"}]}`

jelly commented on .github/workflows/validate-manifest.yml:
```diff
@@ -0,0 +1,31 @@
+        uses: sourcemeta/jsonschema@v16.0.0
+
+      - name: Validate Cockpit manifests
+        run: jsonschema validate doc/modules/ROOT/attachments/schema.json pkg/ --continue
```
```
[jelle@carbon][~/projects/cockpit/cockpit-verify-manifests]%jsonschema validate doc/modules/ROOT/attachments/schema.json pkg/ --continue
/usr/bin/jsonschema:5: DeprecationWarning: The jsonschema CLI is deprecated and will be removed in a future version. Please use check-jsonschema instead, which can be installed from https://pypi.org/project/check-jsonschema/
  from jsonschema.cli import main
usage: jsonschema [-h] [-i INSTANCES] [-F ERROR_FORMAT] [-o {plain,pretty}] [-V VALIDATOR] [--base-uri BASE_URI] [--version] schema
jsonschema: error: unrecognized arguments: doc/modules/ROOT/attachments/schema.json pkg/ --continue
```

jelly replied:
Ok, since when is python-jsonschema no longer hip and why do we need something not even in the Fedora repos?

Venefilyn replied:
As you saw the CLI is deprecated. Their recommendation to use `check-jsonschema` isn't packaged in Fedora either. According to the spec tests it also [fails a few](https://bowtie.report/#/implementations/python-jsonschema) which made me think lets just go with something else. Found this which is also well-maintained https://github.com/sourcemeta/jsonschema/
```
❯ sudo dnf5 search jsonschema
Updating and loading repositories:
Repositories loaded.
Matched fields: name, summary
 python-sphinx-jsonschema-doc.noarch	Documentation for sphinx-jsonschema
 python3-jsonschema+format.noarch	Metapackage for python3-jsonschema: format extras
 python3-jsonschema+format-nongpl.noarch	Metapackage for python3-jsonschema: format-nongpl extras
 python3-jsonschema-path.noarch	Object-oriented JSONSchema
Matched fields: name
 golang-github-xeipuuv-gojsonschema-devel.noarch	Implementation of JSON Schema, draft v4 v6 & v7
 python3-fastjsonschema.noarch	Fastest Python implementation of JSON schema
 python3-jsonschema.noarch	Implementation of JSON Schema validation for Python
 python3-jsonschema-specifications.noarch	JSON Schema meta-schemas and vocabularies, exposed as a Registry
 python3-jsonschema-specifications-doc.noarch	Documentation for the JSON Schema specifications
 python3-jsonschema-specifications-tests.noarch	Tests for the JSON Schema specifications
 python3-sphinx-jsonschema.noarch	Sphinx extension to display JSON Schema
Matched fields: summary
 dt-schema.noarch	Tooling for devicetree validation using YAML and jsonschema
 rust-schemars_derive+default-devel.noarch	Macros for #[derive(JsonSchema)], for use with schemars
 rust-schemars_derive-devel.noarch	Macros for #[derive(JsonSchema)], for use with schemars
 rust-schemars_derive0.8+default-devel.noarch	Macros for #[derive(JsonSchema)], for use with schemars
 rust-schemars_derive0.8-devel.noarch	Macros for #[derive(JsonSchema)], for use with schemars
```

If you want we could go with the Fedora one but I don't want to depend on something with a deprecated CLI.

jelly commented on doc/modules/ROOT/attachments/schema.json:
```diff
@@ -0,0 +1,229 @@
+                  "weight": {
+                    "description": "How much keywords are prioritized over others",
+                    "default": 3,
+                    "type": "number"
```
Should probably be integer?

Venefilyn replied:
Reading the code it seems like floating point is okay but we should have it integer IMO. You're right

jelly replied:
This is not well defined yeah, but I think it makes sense to not allow 10.1.

jelly commented on doc/modules/ROOT/attachments/schema.json:
```diff
@@ -0,0 +1,229 @@
+    },
+    "priority": {
+      "description": "An optional number that specifies which package is preferred in cases where there are conflicts. For example given two packages with the same `name` a package is chosen based on its priority",
+      "type": "number"
```
Should be an integer

jelly commented on doc/modules/ROOT/attachments/schema.json:
```diff
@@ -0,0 +1,229 @@
+            },
+            "order": {
+              "description": "An optional order number to place this menu item or tool. Lower numbers are listed first",
+              "type": "number"
```
Should be an integer

jelly commented on doc/modules/ROOT/attachments/schema.json:
```diff
@@ -0,0 +1,229 @@
+      "type": "string"
+    },
+    "preload": {
+      "description": "A list of identifiers of the components that should be preloaded. Normally, the files of a component are loaded when the user navigates to it for the first time. The files of a preloaded component are loaded immediately after the user logs in, and the initialization code of the component is invoked. \n\nThe value of this field is an array of strings, where each string is one ...
```
Copy pasta fail, `tool` should be `tools`. Please also fix the docs.

jelly commented on doc/modules/ROOT/attachments/schema.json:
```diff
@@ -0,0 +1,229 @@
+    },
+    "version": {
+      "description": "An informational version number for the package",
+      "type": "string"
```
Can also be an integer:

```
[jelle@carbon][~/projects/cockpit/jsonschema]%check-jsonschema --schemafile doc/modules/ROOT/attachments/schema.json examples/**/manifest.json
Schema validation errors were encountered.
  examples/gtk-broadway/manifest.json::$.version: 0 is not of type 'string'
  examples/long-running-process/manifest.json::$.version: 0 is not of type 'string'
  examples/pinger/manifest.json::$.version: 0 is not of type 'string'
  examples/xhr-proxy/manifest.json::$.version: 0 is not of type 'string'
  examples/zoner/manifest.json::$.version: 0 is not of type 'string'
``

Venefilyn replied:
Aha didn't know we had manifest files there. Will add to the test suite as well

jelly commented:
Another issue seems to be that there are no required keys:
```
[jelle@carbon][~/projects/cockpit/jsonschema]%cat -p foo.json
{
}
[jelle@carbon][~/projects/cockpit/jsonschema]%check-jsonschema --schemafile doc/modules/ROOT/attachments/schema.json foo.json
ok -- validation done
```

Hmm this does not seem required, from reading the bridge implementation.

Venefilyn commented:
@jelly I believe all your comments have been addressed if you wanna give it another review

mvollmer reviewed (commented):
No objections. :-)

mvollmer reviewed (approved):
Thanks! This looks good to me, and we can continue to tweak the schema when we find issues.
